Pluggable Google credential mode (pem, json, token) and update all DAOs

rtitle opened the pull request:
Implements pluggable Google credential modes in all Google DAOs. Currently supported:
1. PEM key
2. JSON key
3. Access token

This breaks backwards compatibility so requires a major version bump.

@jmthibault79 FYI

**PR checklist**
- [x] For each library you've modified here, decide whether it requires a major, minor, or no version bump. (Click [here](/broadinstitute/workbench-libs/blob/develop/CONTRIBUTING.md) for guidance)

If you're doing a **major** or **minor** version bump to any libraries:
- [x] Bump the version in `project/Settings.scala` `createVersion()`
- [x] Update `CHANGELOG.md` for those libraries
- [x] I promise I used `@deprecated` instead of deleting code where possible

In all cases:
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green (CI _and_ coverage tests)
- [ ] Squash commits and **merge to develop**
- [ ] Delete branch after merge

After merging, _even if you haven't bumped the version_: 
- [ ] Update `README.md` and the `CHANGELOG.md` for any libs you changed with the new dependency string. The git hash is the same short version you see in GitHub (i.e. seven characters). You can commit these changes straight to develop.

rtitle committed:
Pluggable Google credential mode (pem, json, token) and update all DAOs

rtitle committed:
Get automation to compile

rtitle committed:
Major version bump

rtitle commented:
Gonna look into 2 changes based on comments so far (mostly reminding myself before I go to bed):
1. Make the token mode take an optional callback so it can refresh tokens if they expire. The token mode is really only used by automation tests which _shouldn't_ take more than 1 hour, but still nice to have.
2. Leave the old constructor signatures for backwards compatibility (they should be able to trivially call the new constructor).

jmthibault79 commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleBigQueryDAO.scala:
```diff
@@ -1,38 +1,36 @@
   }
 
-  override def startQuery(accessToken: String, project: GoogleProject, querySql: String): Future[JobReference] = {
+  override def scopes = Seq(BigqueryScopes.BIGQUERY)
```
Nit: this is a `val` everywhere else.

jmthibault79 commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/GoogleCredentialModes.scala:
```diff
@@ -0,0 +1,60 @@
+  /**
+    * Gets a GoogleCredential from a pem file.
+    */
+  case class Pem(serviceAccountClientId: String, pemFile: File, serviceAccountUser: Option[String] = None) extends GoogleCredentialMode {
```
I would like to see `WorkbenchEmail` used for both the SA ID and User here

rtitle replied:
👍  good call

jmthibault79 commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleBigQueryDAO.scala:
```diff
@@ -1,38 +1,36 @@
-  private def bigquery(accessToken: String): Bigquery = {
-    val credential = new GoogleCredential().setAccessToken(accessToken)
-    new Bigquery.Builder(httpTransport, jsonFactory, credential).setApplicationName(appName).build()
+  private lazy val bigquery: Bigquery = {
```
May be out of scope, but can we standardize the pattern we use for Google service creation inside the HTTP DAOs?  No real preference except for including them at the top like this, rather than at the bottom like PubSub.

rtitle replied:
Yeah agreed - I like them at the top too and `lazy val`s. I can update the DAOs to follow this pattern.

jmthibault79 commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleDirectoryDAO.scala:
```diff
@@ -160,18 +141,7 @@ class HttpGoogleDirectoryDAO(serviceAccountClientId: String,
-    new Directory.Builder(httpTransport, jsonFactory, getGroupServiceAccountCredential).setApplicationName(appName).build()
-  }
-
-  private def getGroupServiceAccountCredential: Credential = {
```
exciting to see this go away

jmthibault79 reviewed (approved):
Love it!  with minor changes

rtitle committed:
PR feedback, and keep default constructors for compatiblity

rtitle committed:
Post rebase fixes

rtitle commented:
@dvoet @helgridly maybe you could take a look at this? Thanks

helgridly commented on google/src/main/scala/org/broadinstitute/dsde/workbench/google/HttpGoogleDirectoryDAO.scala:
```diff
@@ -1,32 +1,41 @@
+                            (implicit system: ActorSystem, executionContext: ExecutionContext)
+  extends AbstractHttpGoogleDAO(appName, googleCredentialMode, workbenchMetricBaseName) with GoogleDirectoryDAO {
+
+  @deprecated(message = "This way of instantiating HttpGoogleDirectoryDAO has been deprecated. Please upgrade your configs appropriately.", since = "1.0")
```
it'll still work though, right? might want to say that to avoid enormous alarm

helgridly replied:
also, is it correct to say "upgrade your configs"? is it configs that need updating, or the code that crates the dao?

jmthibault79 replied:
Deprecated implies "still works" to me

rtitle replied:
Yeah it should still work (and deprecated implies "still works" to me as well).

I copied this from elsewhere - agreed that "upgrade your configs" are not the right words. Will fix.

helgridly commented on project/Settings.scala:
```diff
@@ -72,7 +72,7 @@ object Settings {
     name := "workbench-google",
     libraryDependencies ++= googleDependencies,
-    version := createVersion("0.13"),
+    version := createVersion("1.0"),
```
FYI if bumping to this version of wb-libs will cause deprecation warnings in downstream libraries but all code will still compile and work without changes, you can get away with just a minor version bump :)

rtitle replied:
Yeah was thinking that too. 👍

helgridly reviewed (approved):
looks good! i still think a token-refreshing callback might be useful, though. but perhaps that's for another time

rtitle commented:
Thanks @helgridly ! A token-refreshing callback seemed a little tricky to me - the DAOs would need to catch all 401 (I think?) errors and invoke the token-refreshing callback. It might be a nice feature but should maybe be done separately.

helgridly commented:
If you accept something that looks more like a `UserInfo` -- i.e. something that defines its own expiry time -- then you can use that to decide when to refresh.

rtitle commented:
Ah true - that would be simpler.

rtitle committed:
PR feedback:
- fix deprecation language
- 1.0 -> 0.14
- Token mode provides a way to refresh tokens when they expire

rtitle committed:
Post rebase fix

rtitle committed:
Minor fixes

rtitle committed:
Fix compilation error

rtitle committed:
Add a GoogleStorageDAO.objectExists
